azure-migrate/dependencies-at-scale/AzMig_Dependencies.ps1

```diff
@@ -0,0 +1,616 @@
+    }
+
+    $AuthorizationHeader = "Bearer " + $Token
+    $Headers = [ordered]@{Accept = "application/json"; Authorization = $AuthorizationHeader} 
+	
+	if($Environment -eq "AzureUSGovernment") {
+		$baseurl = "https://management.usgovcloudapi.net"
+	}
+	else {
+		$baseurl = "https://management.azure.com" 
+	}
+    
+    return [ordered]@{
+                SubscriptionId = $SubscriptionId
+                Headers = $Headers
+				baseurl = $baseurl
+                }   
+}
+
+function Get-AzMigProject {
+    [CmdletBinding()]
+    Param(
+        [Parameter(Mandatory = $true)][string]$ResourceGroupName,
+        [Parameter(Mandatory = $true)][string]$ProjectName
+    )
+
+    $query = "resources | where type == 'microsoft.migrate/migrateprojects' and resourceGroup == '$ResourceGroupName' and name == '$ProjectName'"
+    $result = $null
+    $result = Search-AzGraph -Query $query
+    if (-not $result) {
+        throw "Project not found"
+    }
+    return $result.Id
+}
+
+function Get-FilteredMachines {
+    [CmdletBinding()]
+    Param(
+        [Parameter(Mandatory = $true)][string]$SiteId,
+        [Parameter(Mandatory = $true)][string]$appliancename,
+        [Parameter()][Hashtable]$Filter,
+        [Parameter(Mandatory = $true)][string]$OutputCsvFile
+    )
+
+    $filterquery=""
+    if($Filter){   
+        foreach($Key in $Filter.Keys){
+           $Val=$Filter[$Key]
+            if ($Key -eq "IPAddresses") {
+                $iprange = "$Val"
+                function CheckIPAddress($address) {
+                    try {
+                        $ip = [System.Net.IPAddress]::Parse($address.Split('/')[0])
+                        if ($ip.AddressFamily -eq [System.Net.Sockets.AddressFamily]::InterNetwork) {
+                           return "IPv4"
+                        }
+                        elseif ($ip.AddressFamily -eq [System.Net.Sockets.AddressFamily]::InterNetworkV6) {
+                           return "IPv6"
+                        }
+                        else {
+                           return "Invalid"
+                        }
+                    }
+                    catch {
+                        return "Invalid"
+                    }
+                }
+         
+                $ipType = CheckIPAddress($iprange)
+             
+                if ($ipType -eq "IPv4" -or $ipType -eq "IPv6") {
+                    $filterquery += "| mv-expand IPAddress=IPAddresses | extend Iprange = '$iprange' | extend result = " + $ipType.ToLower() + "_compare(tostring(IPAddress),tostring(Iprange)) | where result == '0' | project-away result,Iprange | summarize make_list(IPAddress) by tostring(ServerName),tostring(IPAddresses),tostring(Source),tostring(DependencyStatus),tostring(DependencyErrors),tostring(ErrorTimeStamp),tostring(DependencyStartTime),tostring(OperatingSystem),tostring(PowerStatus),tostring(Appliance),tostring(FriendlyNameOfCredentials),tostring(Tags),tostring(ARMID) | project-away list_IPAddress"
```

Comment:
can you explain how the IPRange is working here? In the filter we should have 2 criteria of the IP. 
1. If single IP is being passed, then exact match.
2. If 2 IPs are being passed, then make a range.